src/basic/stat-util.c

```diff
@@ -25,43 +25,131 @@
 #include "stat-util.h"
 #include "string-util.h"
 
-int is_symlink(const char *path) {
-        struct stat info;
+static int verify_stat_at(
+                int fd,
+                const char *path,
+                bool follow,
+                int (*verify_func)(const struct stat *st),
+                bool verify) {
 
-        assert(path);
+        struct stat st;
 
-        if (lstat(path, &info) < 0)
+        assert(fd >= 0 || fd == AT_FDCWD);
+        assert(!isempty(path) || !follow);
+        assert(verify_func);
+
+        if (fstatat(fd, strempty(path), &st,
+                    (isempty(path) ? AT_EMPTY_PATH : 0) | (follow ? 0 : AT_SYMLINK_NOFOLLOW)) < 0)
                 return -errno;
 
-        return !!S_ISLNK(info.st_mode);
+        if (!verify)
+                return verify_func(&st) >= 0;
+
+        return verify_func(&st);
```

Comment:
i'd write this as:

```c
r = verify_func(&st);

return verify ? r : r >= 0;
```

seems easier to read to me, and emphasizes that the difference really is just in the interpretation of the return value